lib/std/Thread/Once.zig

```diff
@@ -0,0 +1,173 @@
+const std = @import("../std.zig");
+const assert = std.debug.assert;
+const testing = std.testing;
+const os = std.os;
+const c = std.c;
+
+const builtin = @import("builtin");
+const target = builtin.target;
+const single_threaded = builtin.single_threaded;
+
+const SpinWait = @import("SpinWait.zig");
+const Atomic = std.atomic.Atomic;
+const Futex = std.Thread.Futex;
+const Once = @This();
+
+impl: Impl = .{},
+
+pub fn call(self: *Once, comptime f: fn () void) void {
```

Comment:
I strongly oppose to this change, making the function a parameter rather than an initialization-time value makes the usage more complex for no good reason.

If memory bloat is your concern here make `f` a comptime-known field.

Reply:
Thinking of making this `comptime f: anytype, args: anytype` to mimic `Thread.spawn`, the `@call` builtin, and general flexibility.

Passing the function at the call site mimics POSIX (which these primitives are based on) pthread_once + supports context based initialization if people want that. Along with POSIX, this is also the API taken by Rust `std::sync::Once` and C++ `std::once_flag`. 

Not sure how complexity is increased in use case. Mind throwing out an example?

Reply:
> Thinking of making this comptime f: anytype, args: anytype to mimic Thread.spawn, the @call builtin, and general flexibility.

Tried in #9028 and got some unexpected backslash, hopefully it'll go better this time.

> Passing the function at the call site mimics POSIX (which these primitives are based on) pthread_once + supports context based initialization if people want that.
 
POSIX is not a good example of API design.

> Along with POSIX, this is also the API taken by Rust std::sync::Once and C++ std::once_flag.

But not by Go's `sync.Once`.

> Not sure how complexity is increased in use case.

The caller now has to think of the function to call and ensure all the callsites use the same one. Do you have any use case for "hot swapping" functions?

Reply:
> Tried in #9028 and got some unexpected backslash

Read through the issue a bit. Seems like it's a valid use case? I'm not really too strong on keeping or removing args as a feature of Once.

> POSIX is not a good example of API design.

POSIX in general is not good I agree (e.g. condvar being only flexible blocking primitive requires holding a lock), but passing the function at call site doesn't seem like additional cognitive overhead (I elaborate below)

> Do you have any use case for "hot swapping" functions?

Frankly, not really. The `Once` use cases that I can come up with are in a single invocation spot and all call the same function. Such cases do include saving context of the call like if the function failed. Multi function/argument on same Once instance was supported by OS so my thought was "why not support it too".

Reasoning in the "practical" sense of Once usage, I'm having trouble imagining its use case at multiple call sites. It's generally just used in one function and multiple things call that function - no further "thinking" like you described.

Do you have different usages? @LemonBoy 

P.S. starting to think the sync primitives shouldn't be designed around pthreads. Maybe just only:
* `Lock` for non-atomic-type mutual exclusion
* `RwLock` for non-atomic-type mutual exclusion + cheap reads
* `Signal` (an [event counts](https://www.1024cores.net/home/lock-free-algorithms/eventcounts)) for blocking on generically (can be used to impl `pthread_cond_t`)
* `Semaphore` blocking for producer/consumer workloads + notification allowed to free the datastructure (unlike `Signal`)
* `Once` for lazy init

Reply:
> Read through the issue a bit. Seems like it's a valid use case? I'm not really too strong on keeping or removing args as a feature of Once.

Definitely a valid use case in my book, the main argument against that was that adding the arguments to the callsite made `Once` easier to misuse. I, for one, would love to have this.

> POSIX in general is not good I agree (e.g. condvar being only flexible blocking primitive requires holding a lock), but passing the function at call site doesn't seem like additional cognitive overhead (I elaborate below)

Since we have an easy way of capturing the function pointer to call at compile-time and given the lack of use-cases for having swappable functions (we don't have closures, that's IMO why other languages kept the pthread-like behaviour) I'd push for a general simplification of the API, keeping the `fn` parameter in the constructor.

The mental overhead I was talking about shows when you have a single `Once` cell that has multiple call sites, keeping all of them in sync requires you to audit every single one (been there, done that. Had to track down some silly bug caused by me calling pthread_once with different functions).

But after thinking about this for a while I'm retiring my "strong" opposition, if the new interface ends up being a problem we can always switch back to the previous one.

> but it's probably too late at this point.

Nah, that's the beauty of a pre 1.0 language :^)